streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCacheBytes.java

```diff
@@ -0,0 +1,437 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * An in-memory LRU cache store similar to {@link MemoryLRUCache} but byte-based, not
+ * record based
+ *
+ * @see org.apache.kafka.streams.state.Stores#create(String)
+ */
+public class MemoryLRUCacheBytes {
+    private final long maxCacheSizeBytes;
+    private long currentSizeBytes = 0;
+    private LRUNode head = null;
+    private LRUNode tail = null;
+    private final Map<String, TreeMap<Bytes, LRUNode>> map;
+    private final Map<String, DirtyEntryFlushListener> listeners;
+    private final Map<String, Set<Bytes>> dirtKeys;
+
+
+    public interface DirtyEntryFlushListener {
+        void apply(final List<DirtyEntry> dirty);
+    }
+
+
+    public MemoryLRUCacheBytes(long maxCacheSizeBytes) {
+        this.maxCacheSizeBytes = maxCacheSizeBytes;
+        this.map = new HashMap<>();
+        listeners = new HashMap<>();
+        dirtKeys = new HashMap<>();
+    }
+
+    /**
+     * Add a listener that is called each time an entry is evicted from the cache or an explicit flush is called
+     * @param namespace
+     * @param listener
+     */
+    public synchronized void addDirtyEntryFlushListener(final String namespace, DirtyEntryFlushListener listener) {
+        this.listeners.put(namespace, listener);
+    }
+
+    public synchronized void flush(final String namespace) {
+        final DirtyEntryFlushListener listener = listeners.get(namespace);
+        if (listener == null) {
+            throw new IllegalArgumentException("No listener for namespace " + namespace + " registered with cache");
+        }
+
+        final Set<Bytes> keys = dirtKeys.get(namespace);
+        if (keys == null || keys.isEmpty()) {
+            return;
+        }
+        final List<DirtyEntry> entries  = new ArrayList<>();
+        for (Bytes key : keys) {
+            final MemoryLRUCacheBytesEntry entry = get(namespace, key.get());
+            entries.add(new DirtyEntry(key, entry.value, entry));
+        }
+        listener.apply(entries);
+        keys.clear();
+    }
+
+    private void callListener(final Bytes key, final LRUNode node) {
+        final MemoryLRUCacheBytesEntry entry = node.entry();
+        final DirtyEntryFlushListener listener = listeners.get(node.namespace);
+        if (listener == null) {
+            return;
+        }
+        listener.apply(Collections.singletonList(new DirtyEntry(key, entry.value, entry)));
+    }
+
+    public synchronized MemoryLRUCacheBytesEntry get(final String namespace, byte[] key) {
+        MemoryLRUCacheBytesEntry entry = null;
+        // get map
+        final TreeMap<Bytes, LRUNode> treeMap = this.map.get(namespace);
+        if (treeMap != null) {
```

Comment:
`treeMap` can only be `null` if namespace has invalid value? If yes, should we throw ?